### ISSUE TITLE
Add synthetic gradient training per "Decoupled Neural Interfaces using Synthetic Gradients"

### ISSUE THREAD

**Issue by ihodes** [fchollet/keras#3527]
Cf. http://128.84.21.199/abs/1608.05343

**Comment by llSourcell** [comment]
yes please

**Comment by rarilurelo** [comment]
Hi.
This is my experimental reproduction of "Decoupled Neural Interfaces using Synthetic Gradients"
https://github.com/rarilurelo/tensorflow-synthetic_gradient

There are two problems in my code for using Keras as a complete wrapper.

1. Initial gradient
Initial gradient can be set by Tensorflow(tf.gradient's grad_ys), but K.gradient has no argument for this.
Does anyone know how to set initial gradient by Theano?

2. control dependencies
An M which synthesis the gradients of the lower layer is trained by the true error descending from the upper layer.
This gif on DeepMind's blog is very useful to understand this flow(https://deepmind.com/blog/decoupled-neural-networks-using-synthetic-gradients/#gif-1).
The true error must be computed **before** the layer updated, so I need to use tf.control_dependencies.
I  think it can be implemented by modifying Model: _make_train_function() and Model's initial argument.
However tensorflow and theano have different ways to control dependencies, so it's difficult and will be ugly.

Thanks.

**Comment by vikasverma1077** [comment]
@rarilurelo Initial grads can be set in theano using known_grads option : theano.gradient.grad(cost, wrt, consider_constant=None, disconnected_inputs='raise', add_names=True, known_grads=None, return_disconnected='zero', null_gradients='raise')

**Comment by rishikksh20** [comment]
@rarilurelo and @llSourcell Please put an example of RNN(LSTM) using synthetic gradient, which is no-where at internet. And one question:  how can we updates the weight using synthetic gradient at particular layer as LSTM layers shared weights so updating the weight of one layer can update the weights of other layers?

**Comment by unoexperto** [comment]
So was this implemented in Keras ?

**Comment by sarim-zafar** [comment]
Any update?

**Comment by x0rb0t** [comment]
nobody cares :(

**Comment by carlthome** [comment]
@x0rb0t :(

**Comment by MichaelFomenko** [comment]
https://github.com/mishig25/synthetic-gradients-keras
